File: app/src/main/java/pl/szczodrzynski/edziennik/data/db/modules/login/LoginStore.java
package pl.szczodrzynski.edziennik.data.db.modules.login;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.room.ColumnInfo;
import androidx.room.Entity;

import com.google.gson.JsonElement;
import com.google.gson.JsonNull;
import com.google.gson.JsonObject;

import pl.szczodrzynski.edziennik.data.db.modules.profiles.ProfileFull;

@Entity(tableName = "loginStores",
        primaryKeys = {"loginStoreId"})
public class LoginStore {
    @ColumnInfo(name = "loginStoreId")
    public int id = -1;
    @ColumnInfo(name = "loginStoreType")
    public int type = -1;
    public final static int LOGIN_TYPE_MOBIDZIENNIK = 1;
    public final static int LOGIN_TYPE_LIBRUS = 2;
    public final static int LOGIN_TYPE_VULCAN = 4;
    public final static int LOGIN_TYPE_IUCZNIOWIE = 3;
    public final static int LOGIN_TYPE_DEMO = 20;
    @ColumnInfo(name = "loginStoreData")
    public JsonObject data;

    @ColumnInfo(name = "loginStoreMode")
    public int mode = 0;
    public static final int LOGIN_MODE_LIBRUS_EMAIL = 0;
    public static final int LOGIN_MODE_LIBRUS_SYNERGIA = 1;
    public static final int LOGIN_MODE_LIBRUS_JST = 2;

    public LoginStore(int id, int type, JsonObject data) {
        this.id = id;
        this.type = type;
        this.data = data;
    }

    public static LoginStore fromProfileFull(ProfileFull profileFull) {
        return new LoginStore(profileFull.getLoginStoreId(), profileFull.getLoginStoreType(), profileFull.getLoginStoreData());
    }

    public void copyFrom(Bundle args) {
        for (String key: args.keySet()) {
            Object o = args.get(key);
            if (o instanceof String) {
                putLoginData(key, (String) o);
            }
            else if (o instanceof Integer) {
                putLoginData(key, (Integer) o);
            }
            else if (o instanceof Long) {
                putLoginData(key, (Long) o);
            }
            else if (o instanceof Float) {
                putLoginData(key, (Float) o);
            }
            else if (o instanceof Boolean) {
                putLoginData(key, (Boolean) o);
            }
        }
    }
    public boolean hasLoginData(String key) {
        if (data == null)
            return false;
        return data.has(key);
    }
    @Nullable
    public String getLoginData(String key, @Nullable String defaultValue) {
        if (data == null)
            return defaultValue;
        JsonElement element = data.get(key);
        if (element != null && !(element instanceof JsonNull)) {
            return element.getAsString();
        }
        return defaultValue;
    }
    @Nullable
    public int getLoginData(String key, int defaultValue) {
        if (data == null)
            return defaultValue;
        JsonElement element = data.get(key);
        if (element != null && !(element instanceof JsonNull)) {
            return element.getAsInt();
        }
        return defaultValue;
    }
    @Nullable
    public long getLoginData(String key, long defaultValue) {
        if (data == null)
            return defaultValue;
        JsonElement element = data.get(key);
        if (element != null && !(element instanceof JsonNull)) {
            return element.getAsLong();
        }
        return defaultValue;
    }
    @Nullable
    public float getLoginData(String key, float defaultValue) {
        if (data == null)
            return defaultValue;
        JsonElement element = data.get(key);
        if (element != null && !(element instanceof JsonNull)) {
            return element.getAsFloat();
        }
        return defaultValue;
    }
    public boolean getLoginData(String key, boolean defaultValue) {
        if (data == null)
            return defaultValue;
        JsonElement element = data.get(key);
        if (element != null && !(element instanceof JsonNull)) {
            return element.getAsBoolean();
        }
        return defaultValue;
    }

    public void putLoginData(String key, String value) {
        forceLoginStore();
        data.addProperty(key, value);
    }
    public void putLoginData(String key, int value) {
        forceLoginStore();
        data.addProperty(key, value);
    }
    public void putLoginData(String key, long value) {
        forceLoginStore();
        data.addProperty(key, value);
    }
    public void putLoginData(String key, float value) {
        forceLoginStore();
        data.addProperty(key, value);
    }
    public void putLoginData(String key, boolean value) {
        forceLoginStore();
        data.addProperty(key, value);
    }

    public void removeLoginData(String key) {
        if (data == null)
            return;
        data.remove(key);
    }

    public void clearLoginStore() {
        data = new JsonObject();
    }

    private void forceLoginStore() {
        if (data == null) {
            clearLoginStore();
        }
    }

    public String type() {
        switch (type) {
            case LOGIN_TYPE_MOBIDZIENNIK:
                return "LOGIN_TYPE_MOBIDZIENNIK";
            case LOGIN_TYPE_LIBRUS:
                return "LOGIN_TYPE_LIBRUS";
            case LOGIN_TYPE_IUCZNIOWIE:
                return "LOGIN_TYPE_IDZIENNIK";
            case LOGIN_TYPE_VULCAN:
                return "LOGIN_TYPE_VULCAN";
            case LOGIN_TYPE_DEMO:
                return "LOGIN_TYPE_DEMO";
            default:
                return "unknown";
        }
    }
    public String mode() {
        switch (mode) {
            case LOGIN_MODE_LIBRUS_EMAIL:
                return "LOGIN_MODE_LIBRUS_EMAIL";
            case LOGIN_MODE_LIBRUS_SYNERGIA:
                return "LOGIN_MODE_LIBRUS_SYNERGIA";
            case LOGIN_MODE_LIBRUS_JST:
                return "LOGIN_MODE_LIBRUS_JST";
            default:
                return "unknown";
        }
    }

    @Override
    public String toString() {
        return "LoginStore{" +
                "id=" + id +
                ", type=" + type() +
                ", mode=" + mode() +
                ", data=" + data +
                '}';
    }
}

File: app/src/main/java/pl/szczodrzynski/edziennik/data/db/entity/LoginStore.kt
/*
 * Copyright (c) Kacper Ziubryniewicz 2020-1-6
 */

package pl.szczodrzynski.edziennik.data.db.entity

import android.os.Bundle
import androidx.room.ColumnInfo
import androidx.room.Entity
import com.google.gson.JsonObject
import pl.szczodrzynski.edziennik.data.db.enums.LoginMode
import pl.szczodrzynski.edziennik.data.db.enums.LoginType
import pl.szczodrzynski.edziennik.ext.*

@Entity(tableName = "loginStores", primaryKeys = ["loginStoreId"])
class LoginStore(
        @ColumnInfo(name = "loginStoreId")
        val id: Int,

        @ColumnInfo(name = "loginStoreType")
        val type: LoginType,

        @ColumnInfo(name = "loginStoreMode")
        val mode: LoginMode,

        @ColumnInfo(name = "loginStoreData")
        val data: JsonObject = JsonObject()
) {

    fun hasLoginData(key: String) = data.has(key)
    fun getLoginData(key: String, defaultValue: Boolean) = data.getBoolean(key) ?: defaultValue
    fun getLoginData(key: String, defaultValue: String?) = data.getString(key) ?: defaultValue
    fun getLoginData(key: String, defaultValue: Int) = data.getInt(key) ?: defaultValue
    fun getLoginData(key: String, defaultValue: Long) = data.getLong(key) ?: defaultValue
    fun getLoginData(key: String, defaultValue: Float) = data.getFloat(key) ?: defaultValue
    fun getLoginData(key: String, defaultValue: Char) = data.getChar(key) ?: defaultValue
    fun getLoginData(key: String, defaultValue: JsonObject) = data.getJsonObject(key) ?: defaultValue
    fun putLoginData(key: String, value: Boolean) { data[key] = value }
    fun putLoginData(key: String, value: String?) { data[key] = value }
    fun putLoginData(key: String, value: Number) { data[key] = value }
    fun putLoginData(key: String, value: Char) { data[key] = value }
    fun putLoginData(key: String, value: JsonObject) { data[key] = value }
    fun removeLoginData(key: String) { data.remove(key) }

    fun copyFrom(args: Bundle) {
        for (key in args.keySet()) {
            when (val o = args[key]) {
                is String -> putLoginData(key, o)
                is Int -> putLoginData(key, o)
                is Long -> putLoginData(key, o)
                is Float -> putLoginData(key, o)
                is Boolean -> putLoginData(key, o)
                is Bundle -> putLoginData(key, o.toJsonObject())
            }
        }
    }

    override fun toString(): String {
        return "LoginStore{" +
                "id=" + id +
                ", type=" + type +
                ", mode=" + mode +
                ", data=" + data +
                '}'
    }
}
